### PR TITLE
Secure bootloader support

### PR DESCRIPTION
**Pull request by bugobliterator**
This breaks ground towards adding functional security in Ardupilot. This is first in series of work towards permission system for Ardupilot use to meet regulatory criterions.

Following commands are used to generate and load secure bootloader (Tested with CubeOrange):
```
#generate private key
openssl ecparam -out privatekey.pem -name secp256r1 -genkey

#Build Secure Bootloader
`./Tools/scripts/build_bootloaders.py CubeOrange --secure-key privatekey.pem`

#Build Secure Firmware with unsecure bootloader
`./waf configure --board CubeOrange --secure-key privatekey.pem`

#Build Secure Firmware with embedded secure bootloader
`./waf configure --board CubeOrange --secure-key privatekey.pem --embed-secure-bootloader````

### CODE REVIEW

File: Tools/AP_Bootloader/asm_arm.inc [hunk 1]
```diff
@@ -0,0 +1,2455 @@
+#define DEC_5 4
```
Comment (peterbarker) [inline Tools/AP_Bootloader/asm_arm.inc:1]: Where did this come from?

**Review by peterbarker — COMMENTED** [review]
Any chance f moving all of those new external files (e.g. asm_arm.inc) into a subdirectory?  No git repo available for them?

What are the regulatory goals?  Can ArduPilot even satisfy those regulatory goals while still conforming to the GPLv3's anti-tivoisation clauses?  Possibly for businesses but not consumers?

File: Tools/AP_Bootloader/asm_arm.inc (hunk 1 above)
Comment (bugobliterator) [reply Tools/AP_Bootloader/asm_arm.inc:1]: This is from uECC. https://github.com/kmackay/micro-ecc

**Comment by bugobliterator** [comment]
> What are the regulatory goals? Can ArduPilot even satisfy those regulatory goals while still conforming to the GPLv3's anti-tivoisation clauses? Possibly for businesses but not consumers?
The technical goal is to have authorisation system in place to operate the vehicle. The authorised use in the case of Indian Regulation is checked by the way of looking at signed logs. For that we actually don't need to lock the bootloader, just need to wipe the keys sector when flashing unauthorised firmware.
Even though I believe I can implement it in a way that it's compatible with TiVo clause. Although I had a discussion with @tridge on this issue a while ago, and tridge's recommendation was that TiVo clause doesn't apply to Drones as they are not consumer electronics... In any case we can have a deeper discussion on this during devcall.

**Comment by tridge** [comment]
I'd like to see an explanation of exactly what the user experience will be with this. How is it deployed? How is fw updated? Can users flash a different fw? What language in the regulations does this line up with?
Apart from those higher level things, the PR is not well structured. The new code should be better separated, I'd like to see a strong justification for including asm (how much faster? smaller? does it matter?), not use "SECURE" as the cpp check.

**Pull request closed without merging**